### PR TITLE
D2IQ-71947 - Add a CSI-volume configuration section to the service form

### PR DESCRIPTION
**Pull request by pierrebeitz**
## Testing

1. ```
   npm start
   open http://localhost:4200/#/services/overview/create
   ```
2. add a volume to a single-container app and select the option that says "CSI".
3. play around with the form and the JSON editor - maybe add more volumes.

## Reviewing

i suggest to review commit-by-commit. i added some comments here on places that i render interesting.

## Trade-offs

as the services form works a little different than the jobs form and has multiple sources of truth from an initialization perspective, please don't be too critical on the DRY-perspective.
i anticipate the services form to shrink A LOT if we continue to use the pattern introduced for this small section (it's the same that has been introduced form jobs-with-dependencies). it should also greatly enhance comprehendabilty as we don't need to manipulate intermediate values "by hand" (to see what i'm talking about [click here](https://github.com/dcos/dcos-ui/blob/108e3e218c409d8989858590d4af64e1bd46912f/plugins/jobs/src/js/components/form/helpers/JobParsers.ts#L27-L167) or [here](https://github.com/dcos/dcos-ui/blob/f096d92e99d5e5c23c072dd56aaeac8951b6fa47/plugins/services/src/js/reducers/serviceForm/JSONReducers/Containers.ts#L351-L592). i'm sure we don't need stuff like that at all. it makes it REALLY hard to reason about the output)

i have not written integration tests yet but will add those in the next time. let's please merge this to enable beta-testing.

## Screenshots

![OS 2020-10-02 10-32-03](https://user-images.githubusercontent.com/300861/94903749-89e6a180-049a-11eb-9687-76831ff8267b.png)

### CODE REVIEW

File: src/mocks/handlers.ts [hunk 1]
```diff
@@ -18,6 +22,8 @@ setupWorker(
   rest.get("/mesos/master/state-summary", fx("marathon-1-task/summary")),
   rest.get("/metadata", fx("dcos/metadata")),
   rest.get("/navstar/lashup/key", (_, res, ctx) => res(ctx.json({}))),
+
+  rest.get("/service/marathon/v2/groups", fx("marathon-1-task/groups")),
```
Comment (pierrebeitz) [inline src/mocks/handlers.ts:26]: these changes enable local development on the `serviceCreateForm` while also fixing a memory leak (we were requiring the same fixtures again and again. `require`d stuff stays in memory 😊)

File: src/js/__tests__/__snapshots__/typecheck-test.ts.snap [hunk 2]
```diff
@@ -3313,6 +3226,7 @@ plugins/services/src/js/reducers/serviceForm/JSONReducers/Volumes.ts: error TS26
 plugins/services/src/js/reducers/serviceForm/JSONReducers/Volumes.ts: error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.
 plugins/services/src/js/reducers/serviceForm/JSONReducers/Volumes.ts: error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.
 plugins/services/src/js/reducers/serviceForm/JSONReducers/Volumes.ts: error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.
+plugins/services/src/js/reducers/serviceForm/JSONReducers/Volumes.ts: error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.
```
Comment (pierrebeitz) [inline src/js/__tests__/__snapshots__/typecheck-test.ts.snap:3229]: i've manually checked all the additions in here. some of them just moved within the file and the other ones only appear because the form now has some typing at all.

File: src/js/__tests__/__snapshots__/tslint-test.ts.snap [hunk 3]
```diff
@@ -347,6 +347,11 @@ exports[`tslint snapshot should be up to date 1`] = `
 /plugins/services/src/js/components/forms/NetworkingFormSection.tsx - Shadowed name: 'networks'
 /plugins/services/src/js/components/forms/NetworkingFormSection.tsx - Shadowed name: 'networks'
 /plugins/services/src/js/components/forms/NetworkingFormSection.tsx - don't declare variable selections to return it immediately
+/plugins/services/src/js/components/forms/VolumesFormSection.tsx - Lambdas are forbidden in JSX attributes due to their rendering performance impact
+/plugins/services/src/js/components/forms/VolumesFormSection.tsx - Lambdas are forbidden in JSX attributes due to their rendering performance impact
+/plugins/services/src/js/components/forms/VolumesFormSection.tsx - Lambdas are forbidden in JSX attributes due to their rendering performance impact
+/plugins/services/src/js/components/forms/VolumesFormSection.tsx - Lambdas are forbidden in JSX attributes due to their rendering performance impact
+/plugins/services/src/js/components/forms/VolumesFormSection.tsx - Lambdas are forbidden in JSX attributes due to their rendering performance impact
```
Comment (pierrebeitz) [inline src/js/__tests__/__snapshots__/tslint-test.ts.snap:354]: i currently consider disabling this rule. while it's a good practice to not have lambdas within JSX, so react can cache components for multiple rerenderings, they don't matter too much in most places.

DCOS-UI is mostly slow due to excessive JSON-parsing and the services table being implemented in a way that does not cache very much (it rerenders all the time).

Those lambdas are helping with readability a lot IMO. WDYT @janisz?

File: plugins/services/src/js/reducers/serviceForm/JSONReducers/Volumes.ts [hunk 4]
```diff
@@ -197,11 +209,17 @@ export function JSONParser(state) {
...
       } else if (item.external != null) {
-        setVolumeProp("type", "EXTERNAL");
-        setVolumeProp("name", item.external.name);
-        setVolumeProp("size", item.external.size);
-        setVolumeProp("options", item.external.options);
-        setVolumeProp("provider", item.external.provider);
+        if (item.external.provider.toLowerCase() === "csi") {
+          const data = { ...item, externalCSI: item.external };
+          memo.push(new Transaction(["volumes", index], data));
+          setVolumeProp("type", "EXTERNAL_CSI");
```
Comment (pierrebeitz) [inline plugins/services/src/js/reducers/serviceForm/JSONReducers/Volumes.ts:215]: the 4 lines above are a trick to not interfere with the existing DVDI-external volumes, as handling those in the same path is REALLY tricky. the internal model will store CSI-information at `externalCSI`. we unpack that in line `42`. obviously we have both, the `Parser` and the `JSONReducer` in this file... 🙈

File: plugins/services/src/js/reducers/serviceForm/FormReducers/Volumes.ts [hunk 5]
```diff
@@ -20,14 +39,17 @@ export function FormReducer(state = [], { type, path, value }) {
...
         break;
     }
     return state;
   }
 
   if (joinedPath.includes("volumes") && type === SET) {
     const index = joinedPath.match(/\d+/)[0];
+    if (`volumes.${index}` === joinedPath) {
+      state[index] = value;
+    }
```
Comment (pierrebeitz) [inline plugins/services/src/js/reducers/serviceForm/FormReducers/Volumes.ts:52]: this is the main trick to get rid of having to implement a reducer-path for every single property. we just set the whole volume at once all the time and carefully (partly via type inference) handle object manipulation.

File: plugins/services/src/js/components/modals/CreateServiceModalForm.tsx [hunk 6]
```diff
@@ -307,46 +324,30 @@ class CreateServiceModalForm extends React.Component {
...
-      batch,
-    };
+  handleFormChange = ({ target }) => {
+    if (!target.name) {
+      return;
+    }
 
-    this.setState(newState);
+    const path = target.name.split(".");
+    const value = target.type === "checkbox" ? target.checked : target.value;
```
Comment (pierrebeitz) [inline plugins/services/src/js/components/modals/CreateServiceModalForm.tsx:339]: `event.target.value` works the same as `event.target.getAttribute("name")` in all browser. additionally it widens our interface for manually crafted events. formerly we got native browser-events for each change that has happened on an input. in the CSI-section we now send events explicitly with exactly the content we want.

File: plugins/services/src/js/components/modals/CreateServiceModalForm.tsx [hunk 7]
```diff
@@ -98,7 +100,32 @@ const JSONEditor = React.lazy(
...
+  },
+  {
+    appConfig: {} | null;
+    baseConfig: unknown;
+    batch: Batch;
+    editedFieldPaths: string[];
+    editingFieldPath: null | string;
+    editorWidth?: number;
+    isPod: boolean;
+  }
```
Comment (pierrebeitz) [inline plugins/services/src/js/components/modals/CreateServiceModalForm.tsx:127]: ☝️ some typing to make a little more sure we don't break stuff while refactoring.

File: plugins/services/src/js/components/forms/VolumesFormSection.tsx [hunk 8]
```diff
@@ -198,24 +185,110 @@ class VolumesFormSection extends React.Component<{
...
-  getExternalVolumeConfig(volume, key) {
-    if (volume.type !== "EXTERNAL") {
-      return null;
-    }
+  getExternalCSIVolumeConfig(volume, key) {
+    const updateVolume = (v) => {
+      this.props.onChange({
+        target: { name: `volumes.${key}`, value: { ...volume, ...v } },
+      });
+    };
```
Comment (pierrebeitz) [inline plugins/services/src/js/components/forms/VolumesFormSection.tsx:193]: here we send our "manual events". (i'll refer to that in later comments).

this is one part of the main trick for not having to implement a reducer-statement that's tied to the name of a native input. so we make things really explicit here.

File: plugins/services/src/js/components/forms/NetworkingFormSection.tsx [hunk 9]
```diff
@@ -59,7 +59,6 @@ class NetworkingFormSection extends mixin(StoreMixin) {
     super.componentDidMount?.();
     this.$dcosVersion = dcosVersion$.subscribe(({ hasCalicoNetworking }) => {
       this.setState({ hasCalicoNetworking });
-      console.log("hasCal", hasCalicoNetworking);
```
Comment (pierrebeitz) [inline plugins/services/src/js/components/forms/NetworkingFormSection.tsx:62]: i just had to remove that. ideally this would go on an own PR. please let me know how you feel about changes like this in the context of a feature-PR. @janisz

File: plugins/services/src/js/components/forms/MultiContainerVolumesFormSection.tsx [hunk 10]
```diff
@@ -24,7 +24,7 @@ import { getContainerNameWithIcon } from "../../utils/ServiceConfigDisplayUtil";
 import { FormReducer as volumeMounts } from "../../reducers/serviceForm/MultiContainerVolumes";
 import VolumeConstants from "../../constants/VolumeConstants";
 
-const excludedTypes = ["DSS", "EXTERNAL"];
+const excludedTypes = ["DSS", "EXTERNAL", "EXTERNAL_CSI"];
```
Comment (pierrebeitz) [inline plugins/services/src/js/components/forms/MultiContainerVolumesFormSection.tsx:27]: we don't want the option to appear for pods. i'll add an integration test for that later.

**Review by DominikDary — APPROVED** [review]

**Comment by danielone** [comment]
@pierrebeitz 

I see already merged but perhaps for a next pass 

1)is Block and Mount the type of Access Mode? Is there a selection by default? would the user choose that before or after? does it affect anything else?

2) to be more standard with how going with design system, instead of saying "entry" be explicit with object adding so user doesn't have to look at header which is important once there are entries there. This also gives user clarity on what will be adding for each section

so 
Add Key Value Pair
Add Node Publish Secret (unless this is a key value pair too ? )
Add Volume Context (unless this is a key value pair too ? )

**Comment by pierrebeitz** [comment]
@danielone next pass is definitely possible.

1) it's the "accessType". "mount" is preselected. it affects whether "Filesystem Type" is visible or not.
2) ok, will do.

**Comment by danielone** [comment]
@pierrebeitz thanks.
1. If Mount is preselected, then may be best as first option. right now it comes off as "type of access mode" so may want to just give it a label as well "Access Type"

**Comment by pierrebeitz** [comment]
@danielone addressing your latest requests in https://jira.d2iq.com/browse/D2IQ-72446 👍
